Numbers: Removed the redundant checks of igcd method and simplified it

supreet11agrawal opened the pull request:
In igcd method, various redundant checks are present such as in line 185 a is surely not equal to 1.
Also in line 192, b != 1 as we have checked that earlier in line 179. Many other such redundancies have been removed.
I have done this as advised by @oscargus and ran all the checks.

<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #15839 

#### Brief description of what is fixed or changed
igcd method is simplified and redundant checks have been removed. All the tests are running. Kindly review and tell me if I should change something @oscargus @asmeurer 

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

supreet11agrawal committed:
Numbers: Removed the redundant checks of igcd method and simplified it

In igcd method various redundant checks are present such as in line 185 a is surely not equal to 1.
Also in line 192, b != 1 as we have checked that earlier in line 179. Many other such redundancies have been removed.
I have done this as advised by @oscargus and ran all the checks.

jksuom commented on sympy/core/numbers.py:
```diff
@@ -179,26 +179,16 @@ def igcd(*args):
-    while k < len(args):
-        ok = as_int(args[k])
-        k += 1
+            a = igcd2(a,b)
```
Please add a space after comma.

supreet11agrawal replied:
ok

oscargus commented:
Looks good to me! (But I have no merge access.)

jksuom commented:
Would it make sense to start by applying `to_int` to all arguments?

supreet11agrawal commented:
> Would it make sense to start by applying `to_int` to all arguments?

No, I don't think that will be good as igcd should give an error when input has non-integer arguments

jksuom commented:
`as_int` will raise ValueError for a non-integer argument.

smichr commented:
I like the idea of converting all to ints from the start:

```python
args = [as_int(i) for i in args]
if 1 in args:
    return 1
a = args.pop()
for b in args:
    a = igcd2(a, b if b > 0 else -b) if b else a
```

jksuom commented:
We could also start with

    args = [abs(as_int(i)) for i in args]

supreet11agrawal commented:
Maybe it will work. @jksuom  I think you should try implementing this and check the tests

supreet11agrawal commented:
@asmeurer Can you review this one

smichr commented:
> line won't work bcz we can't change the list

It's no problem to pop from a list. If it bothers you to alter the input argument -- in this case, the routine is so short that I don't think it matters -- then you could just use `ints = [abs(as_int(i)) for i in args]`
```python
ints = [abs(as_int(i)) for i in args]
if 1 in ints:
    return 1  # S.One or 1, not sure in this context
a = ints.pop()
for b in ints:
    a = igcd2(a, b) if b else a
return a  # or S(a), not sure in this context
```

smichr commented:
...or simply,
```python
ints = filter(None, [abs(as_int(i)) for i in args])
a = ints.pop() if ints else 1
return reduce(igcd2, ints, a) if 1 not in ints else 1
```

oscargus commented:
Thanks @asmeurer ! I will probably be cautious when it comes to merging for a while, still think I need a better understanding of the whole project. However, I can definitely help out, especially when it comes to labels, cleaning up in the lists etc.

supreet11agrawal commented:
@oscargus does something needs to be changed?

jksuom commented:
I would prefer the version in [this comment](https://github.com/sympy/sympy/pull/15841#issuecomment-457712884). It is quite readable and follows closely to the original. I think that the return values could be `1` and `a` as they are currently.

supreet11agrawal commented:
> I would prefer the version in [this comment](https://github.com/sympy/sympy/pull/15841#issuecomment-457712884). It is quite readable and follows closely to the original. I think that the return values could be `1` and `a` as they are currently.

Ok, I will do the necessary changes. I am also quite positive that this would work

supreet11agrawal committed:
Cleaned the code for igcd method in core/numbers.py
args_temp is a temporary list which stores args' items and is changed using pop()

smichr committed:
Merge branch 'master' into 15839_igcd
